### DISCUSSION TITLE
PCF857516-bit IO port extender

### DISCUSSION THREAD

**Discussion by pruwait** [arendst/Tasmota#12879] [Ideas]
Tasmota docs consist information about PCF8574 and PCF8574A I2C 8-bit IO port extender.
How can I use PCF857516-bit IO port extender?

Is it working for this device?

#define USE_I2C                                // Add support for I2C
#define USE_PCF8574                            // [I2cDriver2] Enable PCF8574 I/O Expander (I2C addresses 0x20 - 0x26 and 0x39 - 0x3F) (+1k9 code)
// #define USE_PCF8574_SENSOR                   // enable PCF8574 inputs and outputs in SENSOR message
// #define USE_PCF8574_DISPLAYINPUT             // enable PCF8574 inputs display in Web page
// #define USE_PCF8574_MQTTINPUT                // enable MQTT message & rule process on input change detection : stat/%topic%/PCF8574_INP = {"Time":"2021-03-07T16:19:23+01:00","PCF8574-1_INP":{"D1":1}}

**Comment by Jason2866** [comment]
Since the PCF857516 is a different chip it needs a driver in Tasmota. There is none, so it is not supported.

**Reply by sfromis** [reply]
You can get it supported by creating/amending a driver.

**Comment by nccchirag** [comment]
Are there any updates here? 
Also, what is the suggestive development path - separate **xvdr** for PCF8575 or a separate **xsns**?

**Reply by sfromis** [reply]
The existing driver for PCF8574 is https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_xdrv_driver/xdrv_28_pcf8574.ino

It might make sense to amend the existing driver to also support the PCF8575 if differences are small. Or create a parallel driver in the same name hierarchy.

**Reply by nccchirag** [reply]
okay thank you

**Reply by umognog** [reply]
@nccchirag did you get anywhere with the driver changes?

Looking to use a 16 channel relay for Halloween, I may need to bench this product to get things up on time if nobody else has already done it.

Confirmation of no action would also be useful, as I will backlog looking at it.

**Reply by barbudor** [reply]
There is already support for MCP23017 that is 16 bits

**Reply by sfromis** [reply]
Another strategy for multiple relay outputs is the Shift595 support, likely more straightforward to use than IO extenders.

**Reply by umognog** [reply]
> There is already support for MCP23017 that is 16 bits

Looking to use an existing PCB with an esp12f onboard with the pcf8575ts that came with the relay board. Looking at the driver for the 8574 its a few hundred lines including comments, shouldn't take much work to adapt it... I say now...

**Reply by umognog** [reply]
> Another strategy for multiple relay outputs is the Shift595 support, likely more straightforward to use than IO extenders.

@sfromis I may look into this as I have some esp12f sat there and the relay PCB itself does seem driven by a pair of 595's.

Do you know if use_shiftt595 still needs a compiled bin, or is it configurable now?

Looking at the 8575 datasheet, looks like the addresses & max pins need adjusted and that may be all it is to increase it from 8 to 16 IO

**Reply by sfromis** [reply]
Shift595 support is configurable by building a binary including the support, like for this shift register based 16-channel relay board.
https://templates.blakadder.com/ESP12F_Relay_X16

**Reply by nccchirag** [reply]
> @nccchirag did you get anywhere with the driver changes?
> 
> Looking to use a 16 channel relay for Halloween, I may need to bench this product to get things up on time if nobody else has already done it.
> 
> Confirmation of no action would also be useful, as I will backlog looking at it.

nope, sorry ... I'm not able to allocate time to write the driver ... 
you may use multiple PCF8574 alternatively ... I have tested the same & it works fine!
apart from this, below are a couple of findings when I was modifying the source .. although I had to drop it due to time constraint ... you may ignore them if they are not useful to you 

1. custom user override setting required to define the chip type #define IS_PCF8575 under #define USE_PCF8574
2. PCF8575 i2c read returns 2-byte status (for obvious reasons) so you will need to rewrite code to change from uint8_t to uint16_t
3. struct PCF8574 definition changes if its PCF8575 .. also the input & output functions
4. 41 Read address	40 Write address -- PCF8575 I2C registers 

Few references which can be useful
- https://github.com/RobTillaart/PCF8575
- https://github.com/xreef/PCF8575_library

Lastly, for your application you may consider using [Esphome](https://esphome.io/components/pcf8574.html) .. it has a pre-built component with support for PCF8575

**Reply by Gandalph65** [reply]
Hello,
@nccchirag 
nope, sorry ... I'm not able to allocate time to write the driver ...
you may use multiple PCF8574 alternatively ... I have tested the same & it works fine!
apart from this, below are a couple of findings when I was modifying the source .. although I had to drop it due to time constraint ... you may ignore them if they are not useful to you

    custom user override setting required to define the chip type #define IS_PCF8575 under #define USE_PCF8574
    PCF8575 i2c read returns 2-byte status (for obvious reasons) so you will need to rewrite code to change from uint8_t to uint16_t
    struct PCF8574 definition changes if its PCF8575 .. also the input & output functions
    41 Read address 40 Write address -- PCF8575 I2C registers

Few references which can be useful

    https://github.com/RobTillaart/PCF8575
    https://github.com/xreef/PCF8575_library




I understand you dont can allocate time to write code for PCF8575, so i just started to fork and try to upgrade your PCF8574 Driver.
But there some fact i not able to follow:
Datasheet: PCF8574/PCF8575 tell's me there is no diffrence about read/write the IC both types getting same lenght about IC2-Bus ??
And is may there any way to identify type of PCF8574/PCF8575 automaticaly and than provide 8/16 IO Ports ?
I'll gona try to find an way for just define @ compile time will have support for this type of IO Expander or not.
I hope you still read this subject.
best regards
Jo

**Reply by barbudor** [reply]
Trying to understand why spending time in adapting/creating a driver while there are existing drivers and boards that can provide up to 32 relays
Is there any commercial boards using that device and you have no choice ?
Why is that a must to be supported ?

Also adapting the 8574 driver may make it fatter so it would impact the flash size for only limited use case
It it really have to happen, I would push for a separated driver so that only the people interested would add to their build


I think it's the occasion to re-point a disease of open-source projects like Tasmota
Why not supporting 20 different models of temp & hum sensors, 15 models of CO2 sensors ... just because all of them are cheap on Ali ?
There are - most of time - people available to write a piece of code to add some rarely used feature for the simple reason that they think they need it at that point in time and "why not ?"
But then people move to something else and it falls back to always the same short team of volunteers to maintain that code for which there is limited interrest and not having the appropriate hardware to test.
We already faced that a few time in the past in Tasmota with some dead code now in the repository that no-one have time to maintain.

So I would really insist on : is this really mandatory to support PCF8575 while
- You can use 2 PCF8574
- You can use MCP23017 which is 16 I/O in single chip if you don't want 2 chips
- You can use '595 up to 32 relays
- You can find commercially available boatds with ESP + 8..16..32 relays based on the above

When it comes to chips that are used in commercially available products like the power measument chips used by Sonoff, Shelly, all the chinese manufacturers, it makes sense to support them all.

When it comes to DIY, just consider what is more important for you : the software or the hardware ?
If you want Tasmota, design something that is endorsed by Tasmota and for which you can expect continuous support in the future.
If you absolutely want PCF8575, don't cry if it is not supported by Tasmota

Sorry, my mid-week troll

**Reply by umognog** [reply]
I think @barbudor there has been a flood of premade devices across Amazon, eBay and AliExpress using PCF8575 - I know, I bought one and they are incredibly cheap. I think folks are then simply trying to get what they know installed and working.

In the end, I didn't have the time to update either, but instead installed esphome which has built in PCF8575 support, and installed the esphome add-on in home assistant, which was the end goal.

Many others may be in a similar position and I hope this post helps them. If tasmota is not really the end goal, but a means to an end, the easy answer is esphome.

**Reply by Gandalph65** [reply]
OK i found other Page the Datasheets diffrents PCF8574/PCF8575 once 2Byte other 3Byte lenght. Those Datasheets from Texas Instruments a bit confusing. Need to think about an soloution.

**Comment by eduardostern** [comment]
I've made a Tasmota driver in Berry for PCF8575. You can get it here: https://github.com/eduardostern/Tasmota-PCF8575-Berry-Driver
